lib/spack/spack/spec.py

```diff
@@ -2014,6 +2019,10 @@ def to_node_dict(self, hash=ht.dag_hash):
 
         params = syaml.syaml_dict(sorted(v.yaml_entry() for _, v in self.variants.items()))
 
+        for k, v in params.items():
+            if self.variants[k].propagate:
+                params[k] = PropagateValue(params[k], self.variants[k].propagate)
```

Comment:
Do we really want to start using non built-in types in `Spec.to_node_dict()`? I think this is opening for having subtle bugs later, since we use this method to serialize to JSON:
```
>>> s = spack.spec.Spec("zlib ++shared")
>>> s.to_node_dict()
{'name': 'zlib', 'versions': [':'], 'parameters': {'shared': PropagateValue(value=True, propagate=True)}, 'concrete': False}
```

Reply:
@alalazo Yes it could create bugs later, but we need a way of keeping track of the propagation value for variants so they can don't just lose the value. Do you have an advice for a more readable change? I'm a little stuck

Reply:
Why and where you need to keep this value? Is it for separate concretization of build dependencies, where the concretization is sent out to worker processes?

Reply:
@RikkiButler20 Let me take a stab at this, I'll post a PR soon.

Reply:
Thanks, I'd love your input. This has me racking my brain.

Reply:
No worries, should be fixed by https://github.com/spack/spack/pull/40789 Let's see if tests passes. That PR should take care of:

> Variants do not propagate when concretized separately.

Reply:
#40789 has just been merged.